### ISSUE TITLE
[FRAME core] Clawback PoV Weights For Dispatchables

### ISSUE THREAD

**Issue by ggwpez** [paritytech/polkadot-sdk#209]
The PoV benchmarking aims to estimate the worst-case scenario, which is unlikely to occur on average.  
Too large over-estimations hurt block utilization. There is currently no way to retroactively measure the actual used proof size.  
Having such a retroactive measurement in place would enable a multi iteration approach; alternating between dispatch and assessment phase to reach maximum block utilization.

It would therefore be nice to have a way to measure the weight of a batch of dispatchables after their execution. For example in:
- Scheduler Task dispatch
- XCM dispatch
- Transaction dispatch
- MessageQueue processing

Basically in every situation where we need to fit a batch of Dispatchables into a weight limit, and assume that the pre- and post-dispatch PoV weight differ greatly.

**Comment by pgherveou** [comment]
The pallet-contracts team is looking into this issue, as decreasing `PoV` size is high on our list too.
I like the solution described so far, but I have a couple of points I would like to add to the conversation:

- How about exploring a pure Runtime solution so we can ship this with a forkless update?. We could track all storage reads from the runtime in a global Hashset and counter? It would add some extra memory pressure, and be an overestimation compared to getting the TrieCache size, but do you think it's a viable solution?

- in [Eliminating pre-dispatch weight](https://forum.polkadot.network/t/eliminating-pre-dispatch-weight/400), we talk about shifting from pre-dispatch weight towards dynamically metering the PoV. This could take us a step closer to freeing up devs from benchmarking the PoV weight component (see other on-going research to move to dynamic `ref_time` metering [here](https://hackmd.io/@pepyakin/By7783Wo5) mentioned in the post).

One idea discussed in this [thread](https://forum.polkadot.network/t/eliminating-pre-dispatch-weight/400/11) is to use systemic over-estimation, using strategies like imposing max PoV weight of 1/10th of a block's capacity, ensuring we can fill up to 90% of available block space.

Would love to hear your thoughts on this.

**Comment by bkchr** [comment]
> * We could track all storage reads from the runtime in a global Hashset and counter?

This would not include the trie nodes and thus, we would underestimate the size of the PoV.



> * in [Eliminating pre-dispatch weight](https://forum.polkadot.network/t/eliminating-pre-dispatch-weight/400), we talk about shifting from pre-dispatch weight towards dynamically metering the PoV. This could take us a step closer to freeing up devs

This would require to let the node abort the execution if any storage access would use over the maximum pov weight.

**Comment by pgherveou** [comment]
> This would not include the trie nodes and thus, we would underestimate the size of the PoV.

isn't counting key + value size an over-estimation, since trie path are repeated in each keys?

> This would require to let the node abort the execution if any storage access would use over the maximum pov weight.

Can't we just exclude the transaction when this happens? `ref_time` is already protecting us from unlimited resource usage

**Comment by bkchr** [comment]
> isn't counting key + value size an over-estimation, since trie path are repeated in each key?

No. Trie nodes are consisting up to 16 * 32byte hashes and you don't know how many are there in between the root and your value from the runtime.

> Can't we just exclude the transaction when this happens? `ref_time` is already protecting us from unlimited resource usage

It is about the block producer, who will find out about this while building a block and not before like it is right now. You need to abort the transaction, but the block producer still would need to get paid for this as he has done work.

**Comment by pgherveou** [comment]
Make sense, thanks for your answers.

I think the important thing here is being able to correct down the weight after each extrinsic.
I guess the argument being discussed in the forum on `systemic over-estimation`,  is that once we do that, we could then avoid complex benchmark and  Instead, simply set a generous pre-dispatch weight (for instance, 1/10th of block weight), ensuring that blocks are at least 90% full.

**Comment by pgherveou** [comment]
We (pallet-contracts) are really interested in this issue and would love to contribute. @bkchr @skunert let me know how I can collaborate with you on it.

**Comment by skunert** [comment]
I have started looking into this issue some time ago already and my current draft mostly follows what has been sketched earlier in this issue. I plan to start opening PRs for this soon.

Outline of the current draft:
- In cumulus there will be an externalities extension that  has a reference to the proof recorder and is able to return the currently estimated proof size.
- A new host function that returns the currently estimated proof size to the runtime.
- Extension can be passed down to the block-builder on authoring and to the client for block import.
- For block import, recording will be enabled so that a proof recorder is available for the extension.
- Currently looking into the best option for block validation. Initially suggested in this issue was to use the `TrieCache` to get a size, but I would prefer to use the same recorder mechanism that we use in the other places.
- When this extension and host function are available, the runtime can reclaim some storage proof weight by calling this host function and calculating the diff between the benchmarked weight and the actual one.

**Comment by crystalin** [comment]
> 

> * When this extension and host function are available, the runtime can reclaim some storage proof weight by calling this host function and calculating the diff between the benchmarked weight and the actual one.

That is indeed going to be very useful. Having done recently many benchmark, I think some worst case scenario (specially the system.account or locks....) are too far from reality (due to the worst case estimation of the tree branches).

**Comment by 0xJayPi** [comment]
> I have started looking into this issue some time ago already and my current draft mostly follows what has been sketched earlier in this issue. I plan to start opening PRs for this soon.

Following-up on the pending PR, as it's expected to fix [srlabs_findings/issues/276](https://github.com/paritytech-secops/srlabs_findings/issues/276), @skunert please let me know when you have the PR and its expected ETA for merge. Thanks in advance!

**Comment by skunert** [comment]
> > I have started looking into this issue some time ago already and my current draft mostly follows what has been sketched earlier in this issue. I plan to start opening PRs for this soon.
> 
> Following-up on the pending PR, as it's expected to fix [srlabs_findings/issues/276](https://github.com/paritytech-secops/srlabs_findings/issues/276), @skunert please let me know when you have the PR and its expected ETA for merge. Thanks in advance!

The PR is available here #1462, I am currently waiting on some more reviews (I know some are in progress). However, I can sadly not access the issue you linked, so not sure what it is about.

**Comment by skunert** [comment]
I would like to discuss the reclaiming mechanism on the runtime side. I think there are two main scenarios worth discussing.

### Extrinsics
When we apply extrinsics we can just perform a check of the proof size before and after the application. The diff between the values tells us how much proof size was actually consumed. The difference between the consumed proof size and the benchmarked one can be reclaimed.

I have a simple `SignedExtension` that will do the above and perform the reclaiming. You can find it in this draft PR: https://github.com/paritytech/polkadot-sdk/pull/2270

### Hooks
For reclaiming in hooks the situation is a bit more complicated. 
- `on_initialize`: This hook returns the weight that is used by itself plus whatever should be consumed by `on_finalize`. So after this is run, the storage proof size is basically "out of sync" and we can not use the pre and post-dispatch diff technique we use for the extrinsics.
- `on_idle`: Here we could compare pre and post-run storage proof weight and reclaim accordingly. However, users will probably want to maximize their usage of the remaining weight and will want to manually reclaim inside their implementation so they can get the maximum possible work done.
- `on_finalize`: There should be nothing to do here since its weight has been accounted for and nothing should happen after this anyway.

One way to allow easy manual reclaiming would be to provide a cumulus-supplied `WeightMeter` wrapper that has the standard `WeightMeter` functionality but is enhanced by knowledge about  proof reclaiming. A user would instantiate such a `ProofAwareWeightMeter` and it would internally fetch the proof size. User could at any time they wish call a `proof_refund` method to reclaim differences in weight.

### Alternatives
In an ideal world, we could just set the current weight to the proof size directly, without the diff calculation steps. However, this would require a change in how we handle `on_initialize`, since it is returning weights for `on_finalize` too. We would need to refactor this to be able to differentiate between the two.

These are my thoughts about it, but I am not too experienced with frame, so happy to hear further opinions.
cc @ggwpez @pgherveou

**Comment by pgherveou** [comment]
> on_idle: Here we could compare pre and post-run storage proof weight and reclaim accordingly. However, users will probably want to maximize their usage of the remaining weight and will want to manually reclaim inside their implementation so they can get the maximum possible work done.

Agree with that.
As an example, our custom multi-block migration framework on pallet-contracts, run the migration on_idle.
It tries to make as much progress as possible given the remaining weight.

In a world with Clawback PoV, I would probably adjust the logic, and add an extra check, where I would call the client api to get the adjusted PoV, before trying one more time to make progress and eventually break from the loop.

Alternatively, if the client call is pretty cheap, we could also just call it every time we mutate the remaining_weight, to get the PoV component adjusted to it's real value.
https://github.com/paritytech/polkadot-sdk/blob/951bcceba085f1d3f5d022b1c211e5150fe8d2b2/substrate/frame/contracts/src/lib.rs#L410-L427

**Comment by ggwpez** [comment]
I think for the hooks we can for now let the implementer do the metering. The big user-facing part are the Transactions.

>`on_idle` [...] However, users will probably want to maximize their usage of the remaining weight and will want to manually reclaim inside their implementation so they can get the maximum possible work done.

This is fine and i dont see the problem. In this case, "manually reclaiming" would just mean to call the new HostFn after each loop iter. For example:  

```rust
fn on_idle(limit: Weight) -> Weight {
    for item in work {
        item();
        if new_host_call() > limit.proof {
            break;
        }
    }

    return // Whatever we estimate.
}
```
The current `System::BlockWeight` should not be modified by that hook but by the caller: `Executive` who can then consume the correct weight.

With `on_initialize` and `on_finalize` i would do it similar. Just trust the developer for now that they benchmark these well enough. I think the important part is just to have more block utilization via Transaction Proof metering.  
Both hooks are also going to be deprecated soon in favour of unpaired hooks, who dont have this weird entanglement for their weight returns.

**Comment by skunert** [comment]
Okay, so you guys agree that manual reclaiming for the hooks is the way to go.

> This is fine and i dont see the problem. In this case, "manually reclaiming" would just mean to call the new HostFn after each loop iter. 

Yeah something like that, but since we are getting the remaining block weight passed into `on_idle`, we would need multiple host calls. One in the beginning to check the baseline proof size and then after some work done do check how much weight was actually consumed. To make this easier I was proposing the weightmeter, would look something like this:
```rust
fn on_idle(limit: Weight) -> Weight {
  // The naming tbd
  // ProofAwareWeightMeter internally stores the storage limit as of now by calling hf
  let meter = ProofAwareWeightMeter::new(limit);
  for item in work {
        // Some checking logic that the item actually fits into the remaining weight via benchmark
        ...
        item();
        // This would downcorrect the consumed storage weight on the meter based on diff of now vs initialization time.
        meter.refund_proof_size()
   }

   return meter.consumed()
}
``` 

> For the signed extension: I think it must be in a specific order, so that the Weight SE deduces the correct amount?

What scenarios do you have in mind? You mean other extension already modify the proof weight and it could lead to wrong results? Was not on my radar until now.

> We can also hard-code this into executive directly. Not sure what is better.

From what I saw until now, I think its attractive to have this in a SE. It stays nicely modular and uses the same mechanism as the `CheckWeight` SE. But needs to be separate because this whole thing will be parachain only.

**Comment by skunert** [comment]
#3002
